cmd/ephemeral-installer-register-cluster-and-infraenv/main.go

```diff
@@ -0,0 +1,248 @@
+	clientClusterParams := &installer.V2RegisterClusterParams{
+		NewClusterParams: clusterParams,
+	}
+	clusterResult, registerClusterErr := bmInventory.Installer.V2RegisterCluster(ctx, clientClusterParams)
+	if registerClusterErr != nil {
+		log.Fatal(registerClusterErr, "Failed to register cluster with assisted-service")
+	}
+
+	modelsCluster := clusterResult.Payload
+
+	log.Info("Registered cluster with id: " + modelsCluster.ID.String())
+
+	log.Info("Registering infraenv")
+
+	infraEnvParams := controllers.CreateInfraEnvParams(&infraEnv, models.ImageType(Options.ImageTypeISO), pullSecret, modelsCluster.ID, modelsCluster.OpenshiftVersion)
+
+	if len(staticNetworkConfig) > 0 {
+		log.Infof("Added %d nmstateconfigs", len(staticNetworkConfig))
+		infraEnvParams.InfraenvCreateParams.StaticNetworkConfig = staticNetworkConfig
+	}
+
+	clientInfraEnvParams := &installer.RegisterInfraEnvParams{
+		InfraenvCreateParams: infraEnvParams.InfraenvCreateParams,
+	}
+	infraEnvResult, registerInfraEnvErr := bmInventory.Installer.RegisterInfraEnv(ctx, clientInfraEnvParams)
+	if registerInfraEnvErr != nil {
+		log.Fatal(registerInfraEnvErr, "Failed to register infraenv with assisted-service")
+	}
+
+	modelsInfraEnv := infraEnvResult.Payload
+
+	log.Info("Registered infraenv with id: " + modelsInfraEnv.ID.String())
+}
+
+func getReleaseVersion() string {
+	var clusterImageSet hivev1.ClusterImageSet
+	if err := getFileData(Options.ClusterImageSetFile, &clusterImageSet); err != nil {
+		log.Fatal(err.Error())
+	}
+	return clusterImageSet.Spec.ReleaseImage
+}
+
+func getReleaseVersionAndCpuArch(releaseImage string) (version string, cpuArch string) {
+	// releaseImage is in the form: quay.io/openshift-release-dev/ocp-release:4.9.17-x86_64
+	versionCpuArch := strings.Split(releaseImage, ":")[1]
+	// versionCpuArch is 4.9.17-x86_64
+	versionCpuArchSplit := strings.Split(versionCpuArch, "-")
+	if len(versionCpuArchSplit) > 2 {
+		// case: 4.10.0-rc1-x86_64
+		version = strings.Join(versionCpuArchSplit[:2], "-")
+		cpuArch = versionCpuArchSplit[2]
+	} else {
+		// case: 4.10.1-x86_64
+		version = versionCpuArchSplit[0]
+		cpuArch = versionCpuArchSplit[1]
+	}
+	return version, cpuArch
+}
+
+// Read a Yaml file and unmarshal the contents
+func getFileData(filePath string, output interface{}) error {
+
+	contents, err := os.ReadFile(filePath)
+	if err != nil {
+		err = fmt.Errorf("error reading file %s: %w", filePath, err)
+	} else if err = yaml.Unmarshal(contents, output); err != nil {
+		err = fmt.Errorf("error unmarshalling contents of %s: %w", filePath, err)
+	}
+
+	return err
+}
+
+func validateNMStateConfigAndInfraEnv(nmStateConfig aiv1beta1.NMStateConfig, infraEnv aiv1beta1.InfraEnv) error {
+	if len(nmStateConfig.ObjectMeta.Labels) == 0 {
+		return errors.Errorf("nmstateconfig does not have any labels set")
+	}
+
+	if len(infraEnv.Spec.NMStateConfigLabelSelector.MatchLabels) == 0 {
```

Comment:
Same question here. Must an InfraEnv have a matching NMState?

Reply:
Yes. We are requiring there to be at least one NMState for node0.

Reply:
This is because in a fully DHCPed environment we would not have a way to know which address assisted-service is serving at, otherwise.

Reply:
Ack.